src/Lecture2.hs

```diff
@@ -264,7 +290,19 @@ data EvalError
 It returns either a successful evaluation result or an error.
 -}
 eval :: Variables -> Expr -> Either EvalError Int
-eval = error "TODO"
+eval vars expr = innerEval expr
+  where
+    innerEval :: Expr -> Either EvalError Int
+    innerEval (Lit i) = Right i
+    innerEval (Var s) = checkError $ lookup s vars
+      where
+        checkError (Just i) = Right i
+        checkError Nothing = Left (VariableNotFound s)
+    innerEval (Add a b) = add (innerEval a) (innerEval b)
+      where
+        add (Right c) (Right d) = Right (c + d)
+        add (Left e) _ = Left e
+        add _ (Left e) = Left e
```

Comment:
That's a great helper function! 👏🏻 
Once you learn about Monads, you'll be able to reduce this function to a single line ✂️